### PR TITLE
Avoid using legacy test data in most (but not all) of SymbolTests.swift

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: 

## Summary

This is a test only change to phase out _some_ usage of the "LegacyBundle_DoNotUseInNewTests".

After seeing new code that used `makeDocumentationNodeForSymbol(...)` and noticing that it used "LegacyBundle_DoNotUseInNewTests" internally I decided to update those tests and reduce the usage of that legacy fixture. 

While doing this I had to make several correction to most of those tests because the original `makeDocumentationNodeForSymbol(...)` implementation incorrectly skipped most warnings in the documentation, resulting in tests that verify 0 problems when there was in fact 2 or 3 problems with the documentation. 

## Dependencies

None

## Testing

Nothing in particular. This isn't a user-facing change.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] ~Added~Updated tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] ~Updated documentation if necessary~ Not applicable

### CODE REVIEW

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 1]
```diff
@@ -53,20 +53,18 @@ class SymbolTests: XCTestCase {
                   - name: A parameter
                 - Returns: Return value
                 """,
-            articleContent: """
-                # Leading heading is ignored
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:57]: This was incorrect. It's not ignored. There's a warning about this but the old test helper never reported that warning.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 2]
```diff
@@ -53,20 +53,18 @@ class SymbolTests: XCTestCase {
...
-                
+            extensionFileContent: """
                 @Metadata {
                    @DocumentationExtension(mergeBehavior: override)
                 }
                 """
         )
-        XCTAssert(problems.isEmpty)
+        XCTAssertEqual(problems.count, 0, "Unexpected problems: \(problems.map(\.diagnostic.summary).sorted())")
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:62]: This is technically not necessary but makes it nicer to debug these tests in the future if they break because of a new diagnostic.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 3]
```diff
@@ -106,15 +102,18 @@ class SymbolTests: XCTestCase {
...
                 ### Name of a topic
 
-                - ``MyKit``
-                - ``MyKit/MyClass``
+                - ``ModuleName``
+                - ``ModuleName/SomeClass``
 
                 """
         )
-        XCTAssert(problems.isEmpty)
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:114]: This was incorrect. There's should be two warnings about curating the module and curating the container type but the old test helper never reported those warnings.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 4]
```diff
@@ -53,20 +53,18 @@ class SymbolTests: XCTestCase {
...
+        XCTAssertEqual(problems.count, 0, "Unexpected problems: \(problems.map(\.diagnostic.summary).sorted())")
         
         XCTAssertNil(withArticleOverride.abstract,
-                       "The article overrides—and removes—the abstract from the in-source documenation")
+                       "The article overrides—and removes—the abstract from the in-source documentation")
         XCTAssertNil(withArticleOverride.discussion,
-                       "The article overries the discussion.")
+                       "The article overrides the discussion.")
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:67]: This is just fixing typos

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 5]
```diff
@@ -183,21 +196,18 @@ class SymbolTests: XCTestCase {
...
                     ### Name of a topic
 
-                    - ``MyKit``
-                    - ``MyKit/MyClass``
-
+                    - ``ModuleName``
+                    - ``ModuleName/SomeClass``
                     """
             )
-            XCTAssert(problems.isEmpty)
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:200]: This was incorrect. There are two sources of warnings here:
- The redundant `DocumentationExtension(mergeBehavior: append)` raises a warning
- The cyclic curation raises two warnings.

The old test helper never reported these kinds of warnings so the test was incorrectly asserting that there were no warnings.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 6]
```diff
@@ -484,16 +479,14 @@ class SymbolTests: XCTestCase {
...
+            extensionFileContent: nil
         )
         XCTAssertFalse(problems.isEmpty)
         XCTAssertEqual(withRedirectInArticle.redirects, nil)
 
         XCTAssertEqual(problems.first?.diagnostic.identifier, "org.swift.docc.UnsupportedDocCommentDirective")
-        XCTAssertEqual(problems.first?.diagnostic.range?.lowerBound.line, 3)
-        XCTAssertEqual(problems.first?.diagnostic.range?.lowerBound.column, 1)
+        XCTAssertEqual(problems.first?.diagnostic.range?.lowerBound.line, 14)
+        XCTAssertEqual(problems.first?.diagnostic.range?.lowerBound.column, 18)
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:489]: To help test that source location are reported correctly, the new test uses a non-zero row and column for the start of the doc comment. This helps verify that locations are reported in absolute values rather than relative values.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 7]
```diff
@@ -1112,17 +1106,31 @@ class SymbolTests: XCTestCase {
...
+                startOffset: .init(line: 0, character: 0),
+                url: URL(string: "file:///tmp/File.swift")!
+            )
+            
+            // The `guard` statement` below will handle the `nil` case by failing the test and
+            graph.symbols[myFunctionUSR]?.docComment = newDocComment
+            
+            let newGraphData = try JSONEncoder().encode(graph)
+            try newGraphData.write(to: url.appendingPathComponent("mykit-iOS.symbols.json"))
+        }
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:1131]: There was too much in this test that relied on the legacy fixture so rather than fix it I inlined the setup code from the old test helper here.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 8]
```diff
@@ -1309,6 +1315,8 @@ class SymbolTests: XCTestCase {
                 "org.swift.docc.Metadata.InvalidPageColorInDocumentationComment",
                 "org.swift.docc.Metadata.InvalidTitleHeadingInDocumentationComment",
                 "org.swift.docc.Metadata.InvalidRedirectedInDocumentationComment",
+                
+                "org.swift.docc.unresolvedResource", // For the "test" asset that doesn't exist.
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:1319]: Like elsewhere, this warning should be there but the old test helper never reported this type of warning.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 9]
```diff
@@ -1506,79 +1514,78 @@ class SymbolTests: XCTestCase {
 
     // MARK: - Helpers
     
-    func makeDocumentationNodeForSymbol(
+    private func makeDocumentationNodeForSymbol(
         docComment: String,
-        docCommentLineOffset: Int = 0,
-        articleContent: String?,
+        docCommentLineStart: Int = 11, // an arbitrary non-zero start line
+        extensionFileContent: String?,
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:1563]: I renamed this argument because this parameter doesn't represent the content of an _article_ it represents the content of an _extension file_. The original test helper was confusing in this regard.

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 10]
```diff
@@ -1506,79 +1514,78 @@ class SymbolTests: XCTestCase {
 
     // MARK: - Helpers
     
-    func makeDocumentationNodeForSymbol(
+    private func makeDocumentationNodeForSymbol(
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:1560]: There is enough special knowledge in this test file about the specifics of this test data (for example that the symbol is a function, that it has a "name" parameter, and that it has a return value) for this to be useful in other test files. To prevent other test files from accidentally using it, I made it private (same with the other helper below)

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift (hunk 10 above)
Comment (heckj) [reply Tests/SwiftDocCTests/Semantics/SymbolTests.swift:1560]: would it make sense to push that further to `fileprivate`? (are the uses all in this test suite?)

**Review by heckj — COMMENTED** [review]
Nice updates - as a question (not directly related to this PR) - is there a pattern of adding tests to validate warnings and the edge cases for those as they're added? Is that an intention, or kind of a side effect, of this cleanup?

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift (hunk 10 above)
Comment (d-ronnqvist) [reply Tests/SwiftDocCTests/Semantics/SymbolTests.swift:1560]: `private` is more restrictive than `fileprivate` (or equivalent in top-level declarations) but in this case the file only has one XCTestCase class, so it doesn't make any difference in practice.

**Comment by d-ronnqvist** [comment]
> Nice updates - as a question (not directly related to this PR) - is there a pattern of adding tests to validate warnings and the edge cases for those as they're added? Is that an intention, or kind of a side effect, of this cleanup?

I would say that it's a little bit of both.

Based on a quick regex search we have 239 places in 56 files that do some variation of `XCTAssertEqual(context.problems.count, ...)` or `XCTAssertTrue(context.problems.isEmpty)`. Many of these also use `"Unexpected problems: \(problems.map(\.diagnostic.summary).sorted())")` as the assertion failure message so that it's easy to identify unexpected problem without needing to rerun the tests in a debugger.
I like this pattern because it means that any new warning automatically gets checked in a lot of existing tests and the person who adds the warning is forced to evaluate any potential false positives if those any of these test assertions fail.

At the same time, in this specific PR the additional warning validation is largely a side effect of cleaning up the test helper. Because the original test helper was using a legacy fixture that has lots of warnings, the test helper doesn't reported _all_ warnings. Instead it only reported the warnings raised from merging the symbol markup with the extension file markup. This unfortunately doesn't include any "checkers" that run on that markup, doesn't include curation warnings, doesn't include link warnings, doesn't include asset warnings, etc. _for that same markup_. This led to these test asserting and claiming behaviors that were incorrect. By fixing the test helper so that it reported all warnings, I forced the tests to correct these inaccuracies and deal with these warnings (or correct the markup to caused these warnings).

File: Tests/SwiftDocCTests/Semantics/SymbolTests.swift [hunk 11]
```diff
@@ -106,15 +102,18 @@ class SymbolTests: XCTestCase {
...
+                - ``ModuleName``
+                - ``ModuleName/SomeClass``
 
                 """
         )
-        XCTAssert(problems.isEmpty)
+        XCTAssertEqual(problems.map(\.diagnostic.summary), [
+            "Organizing the module 'ModuleName' under 'ModuleName/SomeClass/someMethod(name:)' isn't allowed",
+            "Organizing 'ModuleName/SomeClass' under 'ModuleName/SomeClass/someMethod(name:)' forms a cycle",
+        ])
```
Comment (anferbui) [inline Tests/SwiftDocCTests/Semantics/SymbolTests.swift:113]: Can we use `verifyProblems` here?

Comment (d-ronnqvist) [reply Tests/SwiftDocCTests/Semantics/SymbolTests.swift:113]: In just one of the other tests, I created an inner `verifyProblems` helper to better deal with the tests running in two configurations with different expected problems. That inner helper function isn't available in this test but this test also doesn't have to deal with different expected problems in different configurations. 

Because of this, I prefer to keep this simple test assertion inline.

**Comment by d-ronnqvist** [comment]
@swift-ci please test
